# Income prediction on census data

# Objective: 
To predict whether income exceeds 50K/yr based on census data

Dataset: Adult Data Set

https://archive.ics.uci.edu/ml/machine-learning-databases/adult/adult.data




Variable description:
    
age: continuous

workclass: Private, Self-emp-not-inc, Self-emp-inc, Federal-gov, Local-gov, State-gov, Without-pay, Never-worked.

fnlwgt: continuous.

education: Bachelors, Some-college, 11th, HS-grad, Prof-school, Assoc-acdm, Assoc-voc, 9th, 7th-8th, 12th, Masters, 1st-4th, 10th, Doctorate, 5th-6th, Preschool.

education-num: continuous.

marital-status: Married-civ-spouse, Divorced, Never-married, Separated, Widowed, Married-spouse-absent, Married-AF-spouse.

occupation: Tech-support, Craft-repair, Other-service, Sales, Exec-managerial, Prof-specialty, Handlers-cleaners, Machine-op-inspct, Adm-clerical, Farming-fishing, Transport-moving, Priv-house-serv, Protective-serv, Armed-Forces.

relationship: Wife, Own-child, Husband, Not-in-family, Other-relative, Unmarried.

race: White, Asian-Pac-Islander, Amer-Indian-Eskimo, Other, Black.

sex: Female, Male.

capital-gain: continuous.

capital-loss: continuous.

hours-per-week: continuous.

native-country: United-States, Cambodia, England, Puerto-Rico, Canada, Germany, Outlying-US(Guam-USVI-etc), India, Japan, Greece, South, China, Cuba, Iran, Honduras, Philippines, Italy, Poland, Jamaica, Vietnam, Mexico, Portugal, Ireland, France, Dominican-Republic, Laos, Ecuador, Taiwan, Haiti, Columbia, Hungary, Guatemala, Nicaragua, Scotland, Thailand, Yugoslavia, El-Salvador, Trinadad&Tobago, Peru, Hong, Holand-Netherlands.

class: >50K, <=50K

In [135]:
# Pandas and Numpy libraries
import pandas as pd
import numpy as np

In [136]:
# For preprocessing the data
from sklearn.preprocessing import Imputer
from sklearn import preprocessing

In [137]:
# To split the dataset into train and test datasets
from sklearn.model_selection import train_test_split

In [138]:
# To model the Gaussian Navie Bayes classifier
from sklearn.naive_bayes import GaussianNB

In [139]:
# To calculate the accuracy score of the model
from sklearn.metrics import accuracy_score

In [140]:
adult_df = pd.read_csv('adult.data', header = None, delimiter=' *, *', engine='python')

Load the dataset. Observe that this file has .data extention

For importing the census data, we are using pandas read_csv() method. This method is a very simple and fast method for importing 
data.

We are passing four parameters. The ‘adult.data’ parameter is the file name. The header parameter is for giving details to pandas
that whether the first row of data consists of headers or not. In our dataset, there is no header. So, we are passing None.

The delimiter parameter is for giving the information the delimiter that is separating the data. Here, we are using ‘ , ’ 
delimiter. This delimiter is to show delete the spaces before and after the data values. This is very helpful when there is 
inconsistency in spaces used with data values.

In [141]:
# Print columns in the adult data set
adult_df.columns

Int64Index([0, 1, 2, 3, 4, 5, 6, 7, 8, 9, 10, 11, 12, 13, 14], dtype='int64')

In [142]:
# Adding headers to the dataframe 
adult_df.columns = ['age', 'workclass', 'fnlwgt', 'education', 'education_num', 'marital_status', 'occupation', 'relationship',
                    'race', 'sex', 'capital_gain', 'capital_loss', 'hours_per_week', 'native_country', 'income']

In [143]:
# Number of records(rows) in the dataframe
len(adult_df)

32561

In [144]:
adult_df

age         workclass  fnlwgt     education  education_num  \
0       39         State-gov   77516     Bachelors             13   
1       50  Self-emp-not-inc   83311     Bachelors             13   
2       38           Private  215646       HS-grad              9   
3       53           Private  234721          11th              7   
4       28           Private  338409     Bachelors             13   
5       37           Private  284582       Masters             14   
6       49           Private  160187           9th              5   
7       52  Self-emp-not-inc  209642       HS-grad              9   
8       31           Private   45781       Masters             14   
9       42           Private  159449     Bachelors             13   
10      37           Private  280464  Some-college             10   
11      30         State-gov  141297     Bachelors             13   
12      23           Private  122272     Bachelors             13   
13      32           Private  205019    Assoc-acdm             12   
14      40           Private  121772     Assoc-voc             11   
15      34           Private  245487       7th-8th              4   
16      25  Self-emp-not-inc  176756       HS-grad              9   
17      32           Private  186824       HS-grad              9   
18      38           Private   28887          11th              7   
19      43  Self-emp-not-inc  292175       Masters             14   
20      40           Private  193524     Doctorate             16   
21      54           Private  302146       HS-grad              9   
22      35       Federal-gov   76845           9th              5   
23      43           Private  117037          11th              7   
24      59           Private  109015       HS-grad              9   
25      56         Local-gov  216851     Bachelors             13   
26      19           Private  168294       HS-grad              9   
27      54                 ?  180211  Some-college             10   
28      39           Private  367260       HS-grad              9   
29      49           Private  193366       HS-grad              9   
...    ...               ...     ...           ...            ...   
32531   30                 ?   33811     Bachelors             13   
32532   34           Private  204461     Doctorate             16   
32533   54           Private  337992     Bachelors             13   
32534   37           Private  179137  Some-college             10   
32535   22           Private  325033          12th              8   
32536   34           Private  160216     Bachelors             13   
32537   30           Private  345898       HS-grad              9   
32538   38           Private  139180     Bachelors             13   
32539   71                 ?  287372     Doctorate             16   
32540   45         State-gov  252208       HS-grad              9   
32541   41                 ?  202822       HS-grad              9   
32542   72                 ?  129912       HS-grad              9   
32543   45         Local-gov  119199    Assoc-acdm             12   
32544   31           Private  199655       Masters             14   
32545   39         Local-gov  111499    Assoc-acdm             12   
32546   37           Private  198216    Assoc-acdm             12   
32547   43           Private  260761       HS-grad              9   
32548   65  Self-emp-not-inc   99359   Prof-school             15   
32549   43         State-gov  255835  Some-college             10   
32550   43  Self-emp-not-inc   27242  Some-college             10   
32551   32           Private   34066          10th              6   
32552   43           Private   84661     Assoc-voc             11   
32553   32           Private  116138       Masters             14   
32554   53           Private  321865       Masters             14   
32555   22           Private  310152  Some-college             10   
32556   27           Private  257302    Assoc-acdm             12   
32557

In [145]:
# Handling missing data
# Test whether there is any null value in our dataset or not. We can do this using isnull() method.
adult_df.isnull().sum()

age               0
workclass         0
fnlwgt            0
education         0
education_num     0
marital_status    0
occupation        0
relationship      0
race              0
sex               0
capital_gain      0
capital_loss      0
hours_per_week    0
native_country    0
income            0
dtype: int64

The above output shows that there is no “null” value in our dataset.

Let’s try to test whether any categorical attribute contains a “?” in it or not. At times there exists “?” or ” ” in place of 
missing values. Using the below code snippet we are going to test whether adult_df data frame consists of categorical variables 
with values as “?”.

In [146]:
for i in [1,2,3,4,5,6]:
    print(i,":",i==5)

1 : False
2 : False
3 : False
4 : False
5 : True
6 : False


In [147]:
for value in ['workclass','education','marital_status','occupation','relationship','race','sex','native_country','income']:
    print(value,":", sum(adult_df[value] == '?'))

workclass : 1836
education : 0
marital_status : 0
occupation : 1843
relationship : 0
race : 0
sex : 0
native_country : 583
income : 0


The output of the above code snippet shows that there are 1836 missing values in workclass attribute. 1843 missing values in 
occupation attribute and 583 values in native_country attribute.

# Data preprocessing

For preprocessing, we are going to make a duplicate copy of our original dataframe.We are duplicating adult_df to adult_df_rev 
dataframe. Observe that we have used deep copy while copying. Why?

In [148]:
## Deep copy of adult_df
adult_df_rev = adult_df.copy(deep=True)

In [149]:
dfa=pd.DataFrame([1,2,3,4,5])

In [150]:
dfb=dfa
dfc=dfa.copy()
dfd=dfa.copy(deep=False)

In [151]:
id(dfa)

4916140408

In [152]:
id(dfd)

4916138056

In [153]:
dfa.iloc[0,0] =6
dfb.iloc[1,0] =7
dfc.iloc[3,0] =8
dfd.iloc[2,0] =9

In [154]:
print(dfa)

   0
0  6
1  7
2  9
3  4
4  5


In [155]:
print(dfb)

   0
0  6
1  7
2  9
3  4
4  5


In [156]:
print(dfc)

   0
0  1
1  2
2  3
3  8
4  5


In [157]:
print(dfd)

   0
0  6
1  7
2  9
3  4
4  5


Before doing missing values handling task, we need some summary statistics of our dataframe. For this, we can use describe() 
method. It can be used to generate various summary statistics, excluding NaN values.

In [158]:
adult_df_rev.describe()

age        fnlwgt  education_num  capital_gain  capital_loss  \
count  32561.000000  3.256100e+04   32561.000000  32561.000000  32561.000000   
mean      38.581647  1.897784e+05      10.080679   1077.648844     87.303830   
std       13.640433  1.055500e+05       2.572720   7385.292085    402.960219   
min       17.000000  1.228500e+04       1.000000      0.000000      0.000000   
25%       28.000000  1.178270e+05       9.000000      0.000000      0.000000   
50%       37.000000  1.783560e+05      10.000000      0.000000      0.000000   
75%       48.000000  2.370510e+05      12.000000      0.000000      0.000000   
max       90.000000  1.484705e+06      16.000000  99999.000000   4356.000000   

       hours_per_week  
count    32561.000000  
mean        40.437456  
std         12.347429  
min          1.000000  
25%         40.000000  
50%         40.000000  
75%         45.000000  
max         99.000000

We are passing an “include” parameter with value as “all”, this is used to specify that. we want summary statistics of all the 
attributes.

In [159]:
adult_df_rev.describe(include= 'all')

age workclass        fnlwgt education  education_num  \
count   32561.000000     32561  3.256100e+04     32561   32561.000000   
unique           NaN         9           NaN        16            NaN   
top              NaN   Private           NaN   HS-grad            NaN   
freq             NaN     22696           NaN     10501            NaN   
mean       38.581647       NaN  1.897784e+05       NaN      10.080679   
std        13.640433       NaN  1.055500e+05       NaN       2.572720   
min        17.000000       NaN  1.228500e+04       NaN       1.000000   
25%        28.000000       NaN  1.178270e+05       NaN       9.000000   
50%        37.000000       NaN  1.783560e+05       NaN      10.000000   
75%        48.000000       NaN  2.370510e+05       NaN      12.000000   
max        90.000000       NaN  1.484705e+06       NaN      16.000000   

            marital_status      occupation relationship   race    sex  \
count                32561           32561        32561  32561  32561   
unique                   7              15            6      5      2   
top     Married-civ-spouse  Prof-specialty      Husband  White   Male   
freq                 14976            4140        13193  27816  21790   
mean                   NaN             NaN          NaN    NaN    NaN   
std                    NaN             NaN          NaN    NaN    NaN   
min                    NaN             NaN          NaN    NaN    NaN   
25%                    NaN             NaN          NaN    NaN    NaN   
50%                    NaN             NaN          NaN    NaN    NaN   
75%                    NaN             NaN          NaN    NaN    NaN   
max                    NaN             NaN          NaN    NaN    NaN   

        capital_gain  capital_loss  hours_per_week native_country income  
count   32561.000000  32561.000000    32561.000000          32561  32561  
unique           NaN           NaN             NaN             42      2  
top              NaN           NaN             NaN  United-States  <=50K  
freq             NaN           NaN             NaN          29170  24720  
mean     1077.648844     87.303830       40.437456            NaN    NaN  
std      7385.292085    402.960219       12.347429            NaN    NaN  
min         0.000000      0.000000        1.000000            NaN    NaN  
25%         0.000000      0.000000       40.000000            NaN    NaN  
50%         0.000000      0.000000       40.000000            NaN    NaN  
75%         0.000000      0.000000       45.000000            NaN    NaN  
max     99999.000000   4356.000000       99.000000            NaN    NaN

# Data imputation 

Some of the categorical values have missing values i.e, “?”. We replace the “?” with the above describe methods top row’s value. 
For example, we replace the “?” values of workplace attribute with “Private” value.

In [160]:
for value in ['workclass','education','marital_status','occupation','relationship','race','sex','native_country','income']:
    #adult_df_rev[value].replace(['?'], [adult_df_rev.describe(include='all')[value][2]],inplace='True')
    replaceValue = adult_df_rev.describe(include='all')[value][2]
    adult_df_rev[value][adult_df_rev[value]=='?'] = replaceValue

/Users/z082092/.local/lib/python3.6/site-packages/ipykernel_launcher.py:4: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame

See the caveats in the documentation: http://pandas.pydata.org/pandas-docs/stable/indexing.html#indexing-view-versus-copy
  after removing the cwd from sys.path.


For Naive Bayes, we need to convert all the data values in one format.

We are going to encode all the labels with the value between 0 and n_classes-1. In the present case, it will be 0 and 1.

For implementing this, we are going to use LabelEncoder of scikit learn library.

In [161]:
# Hot Encoding 
le = preprocessing.LabelEncoder()
workclass_cat = le.fit_transform(adult_df.workclass)
education_cat = le.fit_transform(adult_df.education)
marital_cat   = le.fit_transform(adult_df.marital_status)
occupation_cat = le.fit_transform(adult_df.occupation)
relationship_cat = le.fit_transform(adult_df.relationship)
race_cat = le.fit_transform(adult_df.race)
sex_cat = le.fit_transform(adult_df.sex)
native_country_cat = le.fit_transform(adult_df.native_country)

In [162]:
#initialize the encoded categorical columns
adult_df_rev['workclass_cat'] = workclass_cat
adult_df_rev['education_cat'] = education_cat
adult_df_rev['marital_cat'] = marital_cat
adult_df_rev['occupation_cat'] = occupation_cat
adult_df_rev['relationship_cat'] = relationship_cat
adult_df_rev['race_cat'] = race_cat
adult_df_rev['sex_cat'] = sex_cat
adult_df_rev['native_country_cat'] = native_country_cat

In [163]:
adult_df_rev.head()

age         workclass  fnlwgt  education  education_num  \
0   39         State-gov   77516  Bachelors             13   
1   50  Self-emp-not-inc   83311  Bachelors             13   
2   38           Private  215646    HS-grad              9   
3   53           Private  234721       11th              7   
4   28           Private  338409  Bachelors             13   

       marital_status         occupation   relationship   race     sex  ...  \
0       Never-married       Adm-clerical  Not-in-family  White    Male  ...   
1  Married-civ-spouse    Exec-managerial        Husband  White    Male  ...   
2            Divorced  Handlers-cleaners  Not-in-family  White    Male  ...   
3  Married-civ-spouse  Handlers-cleaners        Husband  Black    Male  ...   
4  Married-civ-spouse     Prof-specialty           Wife  Black  Female  ...   

   native_country  income  workclass_cat education_cat marital_cat  \
0   United-States   <=50K              7             9           4   
1   United-States   <=50K              6             9           2   
2   United-States   <=50K              4            11           0   
3   United-States   <=50K              4             1           2   
4            Cuba   <=50K              4             9           2   

   occupation_cat  relationship_cat  race_cat  sex_cat  native_country_cat  
0               1                 1         4        1                  39  
1               4                 0         4        1                  39  
2               6                 1         4        1                  39  
3               6                 0         2        1                  39  
4              10                 5         2        0                   5  

[5 rows x 23 columns]

In [164]:
#drop the old categorical columns from dataframe
dummy_fields = ['workclass','education','marital_status','occupation','relationship','race', 'sex', 'native_country']
adult_df_rev = adult_df_rev.drop(dummy_fields, axis = 1)


Reindex all the columns properly. We have passed the list of column names as a parameter and axis=1 for reindexing the columns.

In [165]:
adult_df_rev = adult_df_rev.reindex_axis(['age', 'workclass_cat', 'fnlwgt', 'education_cat',
                                    'education_num', 'marital_cat', 'occupation_cat',
                                    'relationship_cat', 'race_cat', 'sex_cat', 'capital_gain',
                                    'capital_loss', 'hours_per_week', 'native_country_cat', 
                                    'income'], axis= 1)
adult_df_rev.head(5)

/Users/z082092/.local/lib/python3.6/site-packages/ipykernel_launcher.py:5: FutureWarning: '.reindex_axis' is deprecated and will be removed in a future version. Use '.reindex' instead.
  """


age  workclass_cat  fnlwgt  education_cat  education_num  marital_cat  \
0   39              7   77516              9             13            4   
1   50              6   83311              9             13            2   
2   38              4  215646             11              9            0   
3   53              4  234721              1              7            2   
4   28              4  338409              9             13            2   

   occupation_cat  relationship_cat  race_cat  sex_cat  capital_gain  \
0               1                 1         4        1          2174   
1               4                 0         4        1             0   
2               6                 1         4        1             0   
3               6                 0         2        1             0   
4              10                 5         2        0             0   

   capital_loss  hours_per_week  native_country_cat income  
0             0              40                  39  <=50K  
1             0              13                  39  <=50K  
2             0              40                  39  <=50K  
3             0              40                  39  <=50K  
4             0              40                   5  <=50K

Now we have created multiple categorical columns like “marital_cat”, “race_cat” etc. !

# Data Slicing

In [95]:
# Arrange data into independent variables and dependent variables
X = adult_df_rev.values[:,:14]  ## Features
Y = adult_df_rev.values[:,14]  ## Target

In [96]:
# Split the data into train and test
# Train data size: 70% of original data
# Test data size: 30% of original data
X_train, X_test, Y_train, Y_test = train_test_split(X, Y, test_size = 0.3, random_state = 10)

Implement Gaussian Naive Bayes

In [97]:
clf = GaussianNB()
clf.fit(X_train, Y_train)

GaussianNB(priors=None, var_smoothing=1e-09)

Now GaussianNB classifier is built. The classifier is trained using training data. We can use fit() method for training it. 
After building a classifier, our model is ready to make predictions. We can use predict() method with test set features as its 
parameters.

In [98]:
Y_pred = clf.predict(X_test)

Accuracy of our Gaussian Naive Bayes model

In [166]:
accuracy_score(Y_test, Y_pred, normalize = True)

0.7925069096120381